### DISCUSSION TITLE
Neural network input question (encoder.cc)

### DISCUSSION THREAD

**Discussion by amarchia** [LeelaChessZero/lc0#1774] [Q&A]
I'm researching reinforcement learning and explainability and I'm a little unfamiliar with the exact formatting of the code here for C++. My goal is to write a parser that will let me examine certain inputs that represent certain concepts (such as material advantage). However I've had some issues using some of the tools I've found to take a particular game board that was input into the network and represent that input as something simpler like the most recent board as a FEN notation. Can anyone give me a breakdown of how the neural network input looks in a way that's a little more detailed than the comments in encoder.cc? Any advice or suggestions would be very appreciated. I've looked at the documentation at https://lczero.org/dev/backend/interface/ 

"// All input planes are 64 value vectors, every element of which is either
// 0 or some value, unique for the plane. Therefore, input is defined as
// a bitmask showing where to set the value, and the value itself.

So, it’s a vector of 112 elements. Every element encodes 8×8 plane using InputPlane struct.

InputPlane contains 64 bit field mask which contains the mask of the non-zero elements of 8×8 input matrix, and a float field value which contains value of those non-zero elements."

I am trying to get a better understanding of which vector matches to what values in terms of representing a board as input. I'm not sure if this is the best place to ask for some clarifications/comments on some of the source code here, but it would be very appreciated. As an example would it be possible to take a given input and represent it in FEN notation? I can imagine that it should be possible to convert from these InputPlane structs to a more simplified and standardized notation format for a particular board's state.

**Comment by brianprichardson** [comment]
There is a description of the input in particular at 
https://github.com/LeelaChessZero/lczero-training/blob/master/tf/chunkparser.py#L250
Note that there have been several prior versions.

Currently in Lc0 there are no inputs like material advantage or any other second order chess concepts beyond just the positions of the pieces themselves.

**Reply by amarchia** [reply]
Thanks for your response. It's obviously not super relevant here, but I actually am trying to reproduce some of the work using linear probing and concept activation vectors of reinforcement learning. So I am using an older version of stockfish to do evaluations. I'm mostly just trying to look at the activations of intermediate layers (which I also have working). 

I've gotten most of the rest of this working, my big road block has been building inputs that I can feed into the neural network here. I'm especially trying to figure out what in particular is encoded in each of these bit planes for example. I mostly want to be able to see how to convert from a given input here back into a human readable representation for the board. Understanding which bits represent what (for example which bits represent pawns, or the queens or whatever) is what I'm trying to find good documentation for.

I have been very curious about a few things as well. Doesn't the input require the 7 previous states of the board for example? How does this system generate these states when the game first starts? I'm not as familiar with C as I would like to be, so reading through this hasn't been as clear as I would have liked. I'll be reading through the code you linked to, but hopefully this response clarifies my question a bit if you have any other information.

**Comment by brianprichardson** [comment]
encoder.cc is likely the best, along with the chunkparser.
There is an outdated tool here (from another fork) that would be helpful to study:
https://github.com/kennyfrc/trainingdata-tool

Don't worry too much about the history planes.  There are several command options as to how to set them say from a FEN when not actually playing a game.  The 8 planes is a holdover from the Go game and are not needed at all, although nets trained with something in the history planes typically are a little stronger.

Going forward, suggest joining the Discord and starting in the #help channel as it is the most active resource.
https://discord.gg/B72YNb48

**Reply by amarchia** [reply]
Thanks for the response again. I'll join the discord and see if I can get some help there. I actually have been using https://github.com/DanielUranga/trainingdata-tool/ to generate games already, and the project you linked to looks like a fork of that. My biggest problem is just removing everything but specific board positions from my dataset that match certain criteria. I can pretty easily generate the input using the tools. I'll just have to sit down and go through all the steps in the code mentioned above so I can figure out how to reverse them to go from properly formatted input back to the board position.
